core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -672,24 +681,42 @@ public String toString()
         // This should not be a ConcurrentHashMap because we want to prevent concurrent addition of new consumers during the
         // removal of existing consumers from this map in addDynamicFilters. This ensures that new consumers don't miss filter completion.
         private final Map<DynamicFilterId, List<Consumer<Map<DynamicFilterId, Domain>>>> dynamicFilterConsumers = new HashMap<>();
-        private final long queryStartTime = System.nanoTime();
+        private long queryAttemptStartTime = System.nanoTime();
 
         private DynamicFilterContext(
                 Session session,
                 Set<DynamicFilterId> dynamicFilters,
-                Map<DynamicFilterId, SettableFuture<Void>> lazyDynamicFilters,
+                Set<DynamicFilterId> lazyDynamicFilters,
                 Set<DynamicFilterId> replicatedDynamicFilters)
         {
             this.session = requireNonNull(session, "session is null");
             this.dynamicFilters = requireNonNull(dynamicFilters, "dynamicFilters is null");
-            this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null");
+            this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null").stream()
```

Comment:
This map is no longer threadsafe (as it is being modified). Should it be guarded by a lock?

Reply:
I left it unguarded earlier on the assumption that when `resetForQueryRetry` is called, no other updates would be going on.
I've now updated it to use an AtomicReference to ImmutableMap. Also added some checks to ensure that any updates from previous attempts don't result in modifications once the new attempt has been registered.